streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java

```diff
@@ -169,6 +172,10 @@ private void registerMetrics() {
         iteratorDurationSensor = StateStoreMetrics.iteratorDurationSensor(taskId.toString(), metricsScope, name(), streamsMetrics);
         StateStoreMetrics.addNumOpenIteratorsGauge(taskId.toString(), metricsScope, name(), streamsMetrics,
                 (config, now) -> numOpenIterators.get());
+        StateStoreMetrics.addOldestOpenIteratorGauge(taskId.toString(), metricsScope, name(), streamsMetrics,
+                (config, now) -> openIterators.isEmpty() ? null :
+                    openIterators.stream().mapToLong(MeteredIterator::startTimestamp).min().getAsLong()
```

Comment:
I don't want to over-engineer (given that we can safely assume that the `openIterator` set should be small), but wondering if this is the best implementation?

In the end, we only want to track the create ts, not the iterators themselves. And for create ts we could just maintain a list if longs, and we would `return list.first()` here, and always append to the end of the list when a new iterator is created? Only "remove" would be more expensive, but we could use a sorted tree for the list, and thus remove would be O(log n) not O(n)).

For this case, we also don't need the `MeteredIterator` helper interface.

Thoughts?

Reply:
Yeah, I considered something like this, but I was concerned that `remove` could be a problem, in particular for two Iterators with the same timestamp.

An alternative would be to use a `ConcurrentSkipListMap<KeyValueIterator<?, ?>, Long>`, which would be O(log(n)) for both insert and remove, and eliminate any concerns about duplicate timestamps; whilst still being sorted by timestamp (given the appropriate comparator on construction).

This also solves the problem that this structure should probably be thread-safe, because Iterators may be opened/closed from various Interactive Query threads.

Reply:
I've pushed a commit with the switch to `ConcurrentSkipListSet`, but I can change it again if we come up with something better.